dns_server/src/crawler/core.rs

```diff
@@ -0,0 +1,546 @@
+// Copyright (c) 2023 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::{
+    collections::{btree_map::Entry, BTreeMap, BTreeSet},
+    marker::PhantomData,
+    net::{IpAddr, Ipv4Addr, Ipv6Addr},
+    sync::Arc,
+    time::Duration,
+};
+
+use common::chain::ChainConfig;
+use crypto::random::{seq::IteratorRandom, Rng};
+use logging::log;
+use p2p::{
+    error::P2pError,
+    message::{AnnounceAddrRequest, PeerManagerMessage, PingRequest, PingResponse},
+    net::{
+        default_backend::transport::TransportAddress,
+        types::{ConnectivityEvent, PeerInfo},
+        NetworkingService,
+    },
+    peer_manager::global_ip::IsGlobalIp,
+    types::peer_address::PeerAddress,
+};
+
+/// How many outbound connection attempts can be made per heartbeat
+const MAX_CONNECTS_PER_HEARTBEAT: usize = 25;
+
+/// When the server drops the unreachable node address. Used for negative caching.
+const PURGE_UNREACHABLE_TIME: Duration = Duration::from_secs(3600);
+
+/// When the server drops the unreachable node address that was once reachable. This should take about a month.
+/// Such a long time is useful if the server itself has prolonged connectivity problems.
+const PURGE_REACHABLE_FAIL_COUNT: u32 = 35;
+
+/// Connection state of a potential node address (outbound only)
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+enum AddressState {
+    Disconnected,
+    Connecting,
+    Connected,
+}
+
+/// Additional state of a potential node address
+struct AddressData {
+    /// Connection state
+    state: AddressState,
+
+    /// Last time when the connection state was updated
+    state_updated_at: Duration,
+
+    /// The number of failed connection attempts.
+    /// New connection attempts are made after a progressive backoff time.
+    /// Resets to 0 when an outbound connection to the address is successful.
+    fail_count: u32,
```

Comment:
How about we move this inside `AddressState::Disconnected` and `AddressState::Connecting`?

Reply:
Sure, updated, thanks.